### PR TITLE
Basic select next point tool and mode

### PR DESCRIPTION
**Pull request by MichaelTamaki**
## Description

Adds a mode called "Set Next Point". It shows "ghost dots" from the next stuntsheet and the flows that get to the next dot. The user can connect dots from the currently selected stuntsheet to the next stuntsheet.

A lot of this is a rough draft... I expect a lot of this to change in the coming months! But for an MVP I think this gets us to where we want to be.

TODO: I need to re-do how dots connect to each other. dotLabelIndex isn't effective.

## Pre-PR checklist

- [x] Ran `npm run serve` and:
  - [x] Checked basic functionality
  - [x] Checked that errors are handled
- [x] Ran `npm run lint`
- [x] Ran `npm run test:unit`
- [x] Ran `npm run test:e2e` and ran relevant tests
- [x] Attached reviewers to PR and pinged on Slack/email

## Screenshots/GIFs

Screenshot_2020-10-28 Calchart

### CODE REVIEW

File: src/components/grapher/Grapher.vue [hunk 1]
```diff
@@ -1,7 +1,13 @@
 <template>
   <div class="grapher">
     <svg
-      class="grapher--svg"
+      :class="`grapher--svg ${
```
Comment (rmpowell77) [inline src/components/grapher/Grapher.vue:4]: Do we usually have a space or `-` on class names?  Should this be `grapher--svg-${...`?

**Review by rmpowell77 — APPROVED** [review]

File: src/components/grapher/Grapher.vue (hunk 1 above)
Comment (MichaelTamaki) [reply src/components/grapher/Grapher.vue:4]: You can have multiple different class names for the same element, so in this case there is a class `grapher--svg`, and another class `grapher--svg-next-point`

File: src/components/menu-bottom/MenuBottom.vue [hunk 2]
```diff
@@ -23,39 +24,48 @@
...
+    tool: new ToolSingleDot(),
+    "data-test": "add-rm",
+    disabled: isSetNextPointMode,
+  },
+  {
+    label: "Set Next Point",
+    icon: "link",
+    tool: new ToolSelectNextPoint(),
+    "data-test": "select-next-point",
+    disabled: !isSetNextPointMode,
```
Comment (adhamrait) [inline src/components/menu-bottom/MenuBottom.vue:60]: Should selecting the Set Next Point tool automatically enable SetNextPointMode? Why does there need to be two steps to use this tool

File: src/components/menu-right/SetNextPoint.vue [hunk 3]
```diff
@@ -0,0 +1,54 @@
+<template>
+  <div>
+    <div class="mb-1 field">
+      <b-switch v-model="isSetNextPointMode" data-test="set-next-point--switch"
+        >Set Next Point Mode</b-switch
```
Comment (adhamrait) [inline src/components/menu-right/SetNextPoint.vue:5]: I think this should have a `size="is-small"` so it's the same size as the other components of menu-right. Thoughts?
| small | normal |
|-|-|
|image  |image|

File: src/store/index.ts [hunk 4]
```diff
@@ -34,14 +34,16 @@ export class CalChartState extends Serializable<CalChartState> {
 
   yardlineNumbers = true;
 
-  grapherSvgPanZoom?: SvgPanZoom.Instance;
+  grapherSvgPanZoom: SvgPanZoom.Instance | null = null;
```
Comment (adhamrait) [inline src/store/index.ts:37]: Curious: Why did you change this?

File: src/store/mutations.ts [hunk 5]
```diff
@@ -53,6 +53,60 @@ const mutations: MutationTree<CalChartState> = {
     const currentSS = getSelectedStuntSheet(state);
     currentSS.addDot(dot);
   },
+  syncDotLabelIndices(
+    state,
+    payload: { currentSSDotIndex: number; nextSSDotIndex: number }
+  ): void {
+    // TODO: Design a better way to connect dots between stuntsheets
```
Comment (adhamrait) [inline src/store/mutations.ts:60]: What if there was a sort of linked list linking dots to one another? Like in `StuntSheetDot.ts` there was a `nextDotIndex` that is just the index of the dot that's associated with it in the next stuntsheet. I think it would be easier to assign the next dot that way, and propogating the dot labels would have to be done through the list.

**Comment by kmdurand** [comment]
The screenshot looks like a video game; love it.

File: tests/unit/store/mutations.spec.ts [hunk 6]
```diff
@@ -5,10 +5,52 @@ import StuntSheet from "@/models/StuntSheet";
 import ContInPlace from "@/models/continuity/ContInPlace";
 import ContETFDynamic from "@/models/continuity/ContETFDynamic";
 import ContEven from "@/models/continuity/ContEven";
+import StuntSheetDot from "@/models/StuntSheetDot";
 
 describe("store/mutations", () => {
   let store: Store<CalChartState>;
 
+  describe("syncDotLabelIndices", () => {
+    // TODO: Update and add tests upon updating syncDotLabelIndices
```
Comment (kmdurand) [inline tests/unit/store/mutations.spec.ts:14]: Do we have an issue for this? TODO comments are good for communicating to a reader that something is unfinished, but unless we throw it into the issue tracker, we might lose track of this.

File: tests/unit/components/menu-right/SetNextPoint.spec.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    store.commit("setToolSelected", tool);
+    await setNextPoint.vm.$nextTick();
+    const message = setNextPoint.find('[data-test="set-next-point--message"]');
+    expect(message.exists()).toBe(true);
+    expect(message.text()).toBe(
+      "Select a dot from the current stunt sheet to start from."
+    );
+  });
+
+  it("updates the message upon selecting a dot", async () => {
```
Comment (kmdurand) [inline tests/unit/components/menu-right/SetNextPoint.spec.ts:53]: Pretty comprehensive testing here 👍

File: src/tools/ToolSelectNextPoint.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+  onClick(event: MouseEvent): void {
+    const [x, y] = BaseTool.convertClientCoordinates(event);
+
+    const stuntSheet: StuntSheet =
+      this.currentSSDotIndex === null
+        ? GlobalStore.getters.getSelectedStuntSheet
+        : GlobalStore.getters.getNextStuntSheet;
+    const existingDotIndex = stuntSheet.stuntSheetDots.findIndex(
+      (dot: StuntSheetDot): boolean => {
+        return x === dot.x && y === dot.y;
```
Comment (kmdurand) [inline src/tools/ToolSelectNextPoint.ts:21]: Does `convertClientCoordinates` make it so that this doesn't have to be quite as precise of a click as this looks?

File: src/tools/BaseTool.ts [hunk 9]
```diff
@@ -9,6 +9,13 @@ import { GlobalStore } from "@/store";
  * - https://www.sitepoint.com/how-to-translate-from-dom-to-svg-coordinates-and-back-again/
  */
 export default abstract class BaseTool {
+  /**
+   * Used in ToolSelectNextPoint. Put in BaseTool due to a circular dependency:
```
Comment (kmdurand) [inline src/tools/BaseTool.ts:13]: Sounds challenging -- you don't think there's another way to get around this?

File: src/components/grapher/Grapher.vue [hunk 10]
```diff
@@ -116,7 +149,7 @@ import StuntSheetDot from "@/models/StuntSheetDot";
 import StuntSheet from "@/models/StuntSheet";
 
 /**
- * Renders the field, the dots of the current stunt sheet, and pending dots
+ * Renders the field, thts of the current stunt sheet, and pending dots
```
Comment (kmdurand) [inline src/components/grapher/Grapher.vue:152]: I think this may have been an accident lol

File: src/tools/BaseTool.ts (hunk 9 above)
Comment (kmdurand) [reply src/tools/BaseTool.ts:13]: Would you mind speaking more to this? I'm getting an uneasy feeling from the mention of ToolSingleDot in this dependency chain, and the size of the chain; if there is a painful source of convoluted bugs/problems here, we should probably do whatever we need to to wrench this out.

File: src/store/mutations.ts [hunk 11]
```diff
@@ -53,6 +53,60 @@ const mutations: MutationTree<CalChartState> = {
...
+        const dotWithDotLabelIndex = currentSS.stuntSheetDots.findIndex(
+          (dot) => dot.dotLabelIndex === i
+        );
+        if (dotWithDotLabelIndex === -1) {
+          break;
+        }
+      }
+      currentDot.dotLabelIndex = i;
+    } else {
+      // TODO: Can we make this not affect the next ss's dot's pairing with the next next ss?
```
Comment (kmdurand) [inline src/store/mutations.ts:86]: Same comment regarding TODOs -- do you think we should have an issue to track this, if this is something we want to do?

File: src/store/mutations.ts [hunk 12]
```diff
@@ -53,6 +53,60 @@ const mutations: MutationTree<CalChartState> = {
     const currentSS = getSelectedStuntSheet(state);
     currentSS.addDot(dot);
   },
+  syncDotLabelIndices(
```
Comment (kmdurand) [inline src/store/mutations.ts:56]: This is a pretty large function. Maybe this can be broken up into sensible pieces?

**Comment by MichaelTamaki** [comment]
Closing... I don't like the UX of this. Will redo in a different PR!
